src/qibolab/instruments/qm/controller.py

```diff
@@ -330,14 +330,13 @@ def register_duration_sweeper_pulses(
                 continue
 
             params = args.parameters[operation(pulse)]
-            channel_ids = args.sequence.pulse_channels(pulse.id)
-            channel = self.channels[channel_ids[0]].logical_channel
+            ids = args.sequence.pulse_channels(pulse.id)
             original_pulse = (
                 pulse if params.amplitude_pulse is None else params.amplitude_pulse
             )
             for value in sweeper.values:
                 sweep_pulse = original_pulse.model_copy(update={"duration": value})
-                sweep_op = self.register_pulse(channel, sweep_pulse)
+                sweep_op = self.register_pulse(ids[0], sweep_pulse)
```

Comment:
So, are we deciding that one sweeper applies to one pulse?

To be fair, that's perfectly fine, as it is possible to sweep multiple pulses with parallel sweepers.
However, if we take this decision, we should also change the `Sweeper` object to accept a single pulse or channel, it would be more consistent (alternatively, we can make an additional check just in the QM driver, but we may instead invest in implementing parallel sweepers for all drivers).

> There should be one-- and preferably only one --obvious way to do it.
> https://peps.python.org/pep-0020/

Reply:
> So, are we deciding that one sweeper applies to one pulse?

No, multiple `pulses` (or `channels`) within the same sweeper are still possible. In fact, the Rabi examples posted above are using the same `Sweeper` for all qubits, see for example: https://github.com/qiboteam/qibocal/blob/f9ef97a21d67f7f258f3de342539cb7dcc5af3ad/src/qibocal/protocols/rabi/length.py#L88

In this specific line `ids` is https://github.com/qiboteam/qibolab/blob/d5b7c6f339ee2259923c23ad618ad5c260bf6ba5/src/qibolab/instruments/qm/controller.py#L333

so we would only have issue if the same `pulse.id` is played on multiple channels in the sequence, not if the `Sweeper` has multiple pulses. Reusing pulses on different channels in the same sequence should be rare (I cannot think of any example right now), but even if it happens, I think it won't cause an issue here. What would happen is that the pulse will only be registered as `operation` in only the first `element` in the QM config, but since the waveform and pulse has been defined, I think QM can actually play it on different elements but it raises a warning. I believe I have seen that happening in 0.1.

> To be fair, that's perfectly fine, as it is possible to sweep multiple pulses with parallel sweepers. However, if we take this decision, we should also change the `Sweeper` object to accept a single pulse or channel, it would be more consistent (alternatively, we can make an additional check just in the QM driver, but we may instead invest in implementing parallel sweepers for all drivers).
> 
> > There should be one-- and preferably only one --obvious way to do it.
> > https://peps.python.org/pep-0020/

True, `Sweeper(pulses: list[Pulse])` can always be replaced by multiple `Sweeper(pulse: Pulse)` that play in parallel and is now possible. However, for reasons explained in https://github.com/qiboteam/qibolab/issues/540#issuecomment-2314829052, I would expect the first to be more efficient at least for QM, but potentially other instruments too, so I decided to keep it. It is also a bit quicker to write, since while we only need to pass `range`/`values` once, while the equivalent parallel sweepers would have all the same `values`, even though this is insignificant. That being said, I fully agree with the *preferably only one* so I would be fine with dropping the list (in a different PR), however I would consider this case as slightly trickier as the two ways are not entirely equivalent if there is really a performance difference.

Reply:
> That being said, I fully agree with the _preferably only one_ rule, so I would be fine with dropping the list (in a different PR). However I would consider this case as slightly trickier, because the two ways are not entirely equivalent if there is really a performance difference.

Ok, if you agree, I'd promote this comment to an issue, and qualify it for 0.3. We will keep both the options in 0.2, and observe their usage, while taking some time to consider whether to keep one or both.

0.2 is large enough, and we can't make everything perfect in one shot.

Reply:
And yes, you're right about the part above. I just confused myself.

Reply:
#1023